discovery_engine_core/core/src/engine.rs

```diff
@@ -810,18 +798,29 @@ async fn update_stacks<'a>(
         return Ok(());
     }
 
-    let key_phrases = needy_stacks
+    let key_phrases_by_market = needy_stacks
         .iter()
         .any(|stack| stack.ops.needs_key_phrases())
-        .then(|| ranker.take_key_phrases(market, take_top))
+        .then(|| {
+            markets
+                .iter()
+                .map(|market| (market, ranker.take_key_phrases(market, take_top)))
+                .collect::<HashMap<_, _>>()
+        })
         .unwrap_or_default();
 
     // Here we gather new documents for all relevant stacks, and put them into a vector.
     // We don't update the stacks immediately, because we want to de-duplicate the documents
     // across stacks first.
+    let no_key_phrases = vec![];
     let new_document_futures = needy_stacks
         .iter()
-        .map(|stack| fetch_new_documents_for_stack(stack, ranker, &key_phrases, history, market))
+        .flat_map(|stack| {
+            markets.iter().map(|market| {
+                let key_phrases = key_phrases_by_market.get(market).unwrap_or(&no_key_phrases);
```

Comment:
:art: it would be nice to use `Option::unwrap_or_default()` but there seems no way to get from an `Option<&Vec<_>>` to an `Option<&[_]>` even with `Option::as_deref()`. still there is another possibility which at least allows us to avoid binding the empty `no_key_phrases`:
```suggestion
                let key_phrases = key_phrases_by_market.get(market).map_or_else(<&[_]>::default, Vec::as_slice);
```
or maybe better without the closure
```suggestion
                let key_phrases = key_phrases_by_market.get(market).map_or(&[] as &[_], Vec::as_slice);;
```